d2layouts/d2grid/layout.go

```diff
@@ -477,41 +532,128 @@ func (gd *gridDiagram) getBestLayout(targetSize float64, columns bool) [][]*d2gr
+				newDebt := size - targetSize
+				debt += newDebt
+			} else {
+				rowSize += size
+			}
+			continue
+		}
+		// debt is paid by decreasing threshold to start new row and ending below targetSize
+		if rowSize+(gap+size)/2. > targetSize-debt {
+			fastDivision = append(fastDivision, i-1)
+			newDebt := rowSize - targetSize
+			debt += newDebt
+			rowSize = size
+		} else {
+			rowSize += gap + size
+		}
+	}
+	if len(fastDivision) == nCuts {
+		layout := genLayout(gd.objects, fastDivision)
+		dist := getDistToTarget(layout, targetSize, float64(gd.horizontalGap), float64(gd.verticalGap), columns)
+		if dist < bestDist {
+			bestLayout = layout
+			bestDist = dist
+		}
+	}
 	return bestLayout
 }
 
+// process current division, return true to stop iterating
+type iterDivision func(division []int) (done bool)
+type checkCut func(objects []*d2graph.Object, starting bool) (ok bool)
+
 // get all possible divisions of objects by the number of cuts
-func genDivisions(objects []*d2graph.Object, nCuts int) (divisions [][]int) {
+func iterDivisions(objects []*d2graph.Object, nCuts int, f iterDivision, check checkCut) {
 	if len(objects) < 2 || nCuts == 0 {
-		return nil
+		return
 	}
+	done := false
 	// we go in this order to prefer extra objects in starting rows rather than later ones
 	lastObj := len(objects) - 1
+	// with objects=[A, B, C, D, E]; nCuts=2
+	// d:depth; i:index; n:nCuts;
+	// ┌────┬───┬───┬─────────────────────┬────────────┐
+	// │ d  │ i │ n │ objects             │ cuts       │
+	// ├────┼───┼───┼─────────────────────┼────────────┤
+	// │ 0  │ 4 │ 2 │ [A   B   C   D | E] │            │
+	// ├────┼───┼───┼─────────────────────┼────────────┤
+	// │ └1 │ 3 │ 1 │ [A   B   C | D]     │ + | E]     │
+	// ├────┼───┼───┼─────────────────────┼────────────┤
+	// │ └1 │ 2 │ 1 │ [A   B | C   D]     │ + | E]     │
+	// ├────┼───┼───┼─────────────────────┼────────────┤
+	// │ └1 │ 1 │ 1 │ [A | B   C   D]     │ + | E]     │
+	// ├────┼───┼───┼─────────────────────┼────────────┤
+	// │ 0  │ 3 │ 2 │ [A   B   C | D   E] │            │
+	// ├────┼───┼───┼─────────────────────┼────────────┤
+	// │ └1 │ 2 │ 1 │ [A   B | C]         │ + | D E]   │
+	// ├────┼───┼───┼─────────────────────┼────────────┤
+	// │ └1 │ 1 │ 1 │ [A | B   C]         │ + | D E]   │
+	// ├────┼───┼───┼─────────────────────┼────────────┤
+	// │ 0  │ 2 │ 2 │ [A   B | C   D   E] │            │
+	// ├────┼───┼───┼─────────────────────┼────────────┤
+	// │ └1 │ 1 │ 1 │ [A | B]             │ + | C D E] │
+	// └────┴───┴───┴─────────────────────┴────────────┘
 	for index := lastObj; index >= nCuts; index-- {
+		if !check(objects[index:], false) {
+			// optimization: if current cut gives a bad grouping, don't recurse
+			continue
+		}
 		if nCuts > 1 {
-			for _, inner := range genDivisions(objects[:index], nCuts-1) {
-				divisions = append(divisions, append(inner, index-1))
-			}
+			iterDivisions(objects[:index], nCuts-1, func(inner []int) bool {
+				done = f(append(inner, index-1))
```

Comment:
I wonder if these could end up with those issues of slice references since you're expanding it recursively